File: distance_metrics.py
import numpy as np
from itertools import permutations

# DISTANCE METRICS POUR SÉRIES TEMPORELLES CATÉGORIQUES

# 1. HAMMING DISTANCE (Dynamic Hamming Distance)


def hamming_distance(seq1, seq2):
    """
    Calculer la distance Hamming entre deux séquences
    
    Distance = nombre de positions où les deux séquences diffèrent
    
    Args:
        seq1, seq2: arrays de même longueur
    
    Returns:
        distance: nombre de positions différentes
    """
    if len(seq1) != len(seq2):
        raise ValueError("Les séquences doivent avoir la même longueur")
    
    distance = np.sum(seq1 != seq2)
    return distance


def hamming_distance_matrix(sequences):
    """
    Créer matrice de distances Hamming pour toutes les paires de séquences
    
    Args:
        sequences: array (n_sequences, seq_length)
    
    Returns:
        dist_matrix: array (n_sequences, n_sequences) symétrique
    """
    n_sequences = sequences.shape[0]
    dist_matrix = np.zeros((n_sequences, n_sequences))
    
    for i in range(n_sequences):
        for j in range(i+1, n_sequences):
            distance = hamming_distance(sequences[i], sequences[j])
            dist_matrix[i, j] = distance
            dist_matrix[j, i] = distance
    
    return dist_matrix



# 2. OPTIMAL MATCHING DISTANCE (OM)


def cost_matrix(seq1, seq2, indel_cost=1, substitution_cost=1):
    """
    Créer matrice de coûts pour Optimal Matching
    
    Args:
        seq1, seq2: séquences à comparer
        indel_cost: coût de insertion/deletion
        substitution_cost: coût de substitution
    
    Returns:
        cost_matrix: matrice (len(seq1)+1, len(seq2)+1)
    """
    len1, len2 = len(seq1), len(seq2)
    matrix = np.zeros((len1 + 1, len2 + 1))
    
    # Initialiser première ligne et colonne (coûts d'insertion)
    for i in range(len1 + 1):
        matrix[i, 0] = i * indel_cost
    for j in range(len2 + 1):
        matrix[0, j] = j * indel_cost
    
    # Remplir la matrice avec programmation dynamique
    for i in range(1, len1 + 1):
        for j in range(1, len2 + 1):
            # Coût si on substitue
            if seq1[i-1] == seq2[j-1]:
                substitution = matrix[i-1, j-1]  # Pas de coût si identique
            else:
                substitution = matrix[i-1, j-1] + substitution_cost
            
            # Coût si on delete
            deletion = matrix[i-1, j] + indel_cost
            
            # Coût si on insert
            insertion = matrix[i, j-1] + indel_cost
            
            # Prendre le minimum
            matrix[i, j] = min(substitution, deletion, insertion)
    
    return matrix


def optimal_matching_distance(seq1, seq2, indel_cost=1, substitution_cost=1):
    """
    Calculer Optimal Matching distance entre deux séquences
    
    C'est le coût minimum pour transformer seq1 en seq2
    via insertions, deletions, substitutions
    
    Args:
        seq1, seq2: séquences à comparer
        indel_cost: coût de insertion/deletion
        substitution_cost: coût de substitution
    
    Returns:
        distance: coût minimum
    """
    matrix = cost_matrix(seq1, seq2, indel_cost, substitution_cost)
    return matrix[-1, -1]


def optimal_matching_matrix(sequences, indel_cost=1, substitution_cost=1):
    """
    Créer matrice de distances Optimal Matching pour toutes les paires
    
    Args:
        sequences: array (n_sequences, seq_length)
        indel_cost: coût de insertion/deletion
        substitution_cost: coût de substitution
    
    Returns:
        dist_matrix: array (n_sequences, n_sequences) symétrique
    """
    n_sequences = sequences.shape[0]
    dist_matrix = np.zeros((n_sequences, n_sequences))
    
    for i in range(n_sequences):
        for j in range(i+1, n_sequences):
            distance = optimal_matching_distance(
                sequences[i], 
                sequences[j], 
                indel_cost=indel_cost, 
                substitution_cost=substitution_cost
            )
            dist_matrix[i, j] = distance
            dist_matrix[j, i] = distance
    
    return dist_matrix

# 3. NORMALISER LES DISTANCES


def normalize_distance_matrix(dist_matrix):
    """
    Normaliser la matrice de distances au range [0, 1]
    
    Args:
        dist_matrix: matrice de distances
    
    Returns:
        normalized: matrice normalisée
    """
    max_dist = np.max(dist_matrix)
    if max_dist == 0:
        return dist_matrix
    return dist_matrix / max_dist


# TEST


if __name__ == "__main__":
    # Importer les données générées
    import pandas as pd
    
    sequences = pd.read_csv("sequences_m1.csv").values
    
    print("=" * 60)
    print("CALCUL DES DISTANCE MATRICES")
    print("=" * 60)
    
    # Hamming Distance
    print("\n1. HAMMING DISTANCE")
    print("-" * 60)
    dist_hamming = hamming_distance_matrix(sequences)
    print(f"Matrice Hamming shape: {dist_hamming.shape}")
    print(f"Min distance: {np.min(dist_hamming[np.nonzero(dist_hamming)]):.2f}")
    print(f"Max distance: {np.max(dist_hamming):.2f}")
    print(f"Mean distance: {np.mean(dist_hamming[np.nonzero(dist_hamming)]):.2f}")
    
    # Optimal Matching
    print("\n2. OPTIMAL MATCHING DISTANCE")
    print("-" * 60)
    dist_om = optimal_matching_matrix(sequences)
    print(f"Matrice OM shape: {dist_om.shape}")
    print(f"Min distance: {np.min(dist_om[np.nonzero(dist_om)]):.2f}")
    print(f"Max distance: {np.max(dist_om):.2f}")
    print(f"Mean distance: {np.mean(dist_om[np.nonzero(dist_om)]):.2f}")
    
    # Sauvegarder
    print("\n3. SAUVEGARDE")
    print("-" * 60)
    np.save("distance_matrix_hamming.npy", dist_hamming)
    np.save("distance_matrix_om.npy", dist_om)
    print("✓ Matrices sauvegardées:")
    print("  - distance_matrix_hamming.npy")
    print("  - distance_matrix_om.npy")

File: distance_metrics.R

# DISTANCE METRICS POUR SÉRIES TEMPORELLES CATÉGORIQUES



# 1. HAMMING DISTANCE


hamming_distance <- function(seq1, seq2) {
  """
  Calculate Hamming distance between two sequences
  
  Distance = number of positions where sequences differ
  
  Args:
    seq1, seq2: vectors of same length
  
  Returns:
    distance: number of different positions
  """
  if (length(seq1) != length(seq2)) {
    stop("Sequences must have the same length")
  }
  
  distance <- sum(seq1 != seq2)
  return(distance)
}


hamming_distance_matrix <- function(sequences) {
  """
  Create Hamming distance matrix for all pairs of sequences
  
  Args:
    sequences: matrix (n_sequences, seq_length)
  
  Returns:
    dist_matrix: symmetric matrix (n_sequences, n_sequences)
  """
  n_sequences <- nrow(sequences)
  dist_matrix <- matrix(0, nrow = n_sequences, ncol = n_sequences)
  
  for (i in 1:n_sequences) {
    for (j in i:n_sequences) {
      distance <- hamming_distance(sequences[i, ], sequences[j, ])
      dist_matrix[i, j] <- distance
      dist_matrix[j, i] <- distance
    }
  }
  
  return(dist_matrix)
}



# 2. OPTIMAL MATCHING DISTANCE (OM)


cost_matrix <- function(seq1, seq2, indel_cost = 1, substitution_cost = 1) {
  """
  Create cost matrix for Optimal Matching
  
  Args:
    seq1, seq2: sequences to compare
    indel_cost: cost of insertion/deletion
    substitution_cost: cost of substitution
  
  Returns:
    cost_matrix: matrix (len(seq1)+1, len(seq2)+1)
  """
  len1 <- length(seq1)
  len2 <- length(seq2)
  matrix_cost <- matrix(0, nrow = len1 + 1, ncol = len2 + 1)
  
  # Initialize first row and column (insertion costs)
  for (i in 1:(len1 + 1)) {
    matrix_cost[i, 1] <- (i - 1) * indel_cost
  }
  for (j in 1:(len2 + 1)) {
    matrix_cost[1, j] <- (j - 1) * indel_cost
  }
  
  # Fill matrix with dynamic programming
  for (i in 2:(len1 + 1)) {
    for (j in 2:(len2 + 1)) {
      # Cost if we substitute
      if (seq1[i-1] == seq2[j-1]) {
        substitution <- matrix_cost[i-1, j-1]  # No cost if identical
      } else {
        substitution <- matrix_cost[i-1, j-1] + substitution_cost
      }
      
      # Cost if we delete
      deletion <- matrix_cost[i-1, j] + indel_cost
      
      # Cost if we insert
      insertion <- matrix_cost[i, j-1] + indel_cost
      
      # Take minimum
      matrix_cost[i, j] <- min(substitution, deletion, insertion)
    }
  }
  
  return(matrix_cost)
}


optimal_matching_distance <- function(seq1, seq2, indel_cost = 1, substitution_cost = 1) {
  """
  Calculate Optimal Matching distance between two sequences
  
  This is the minimum cost to transform seq1 into seq2
  via insertions, deletions, substitutions
  
  Args:
    seq1, seq2: sequences to compare
    indel_cost: cost of insertion/deletion
    substitution_cost: cost of substitution
  
  Returns:
    distance: minimum cost
  """
  matrix_cost <- cost_matrix(seq1, seq2, indel_cost, substitution_cost)
  return(matrix_cost[length(seq1) + 1, length(seq2) + 1])
}


optimal_matching_matrix <- function(sequences, indel_cost = 1, substitution_cost = 1) {
  """
  Create Optimal Matching distance matrix for all pairs
  
  Args:
    sequences: matrix (n_sequences, seq_length)
    indel_cost: cost of insertion/deletion
    substitution_cost: cost of substitution
  
  Returns:
    dist_matrix: symmetric matrix (n_sequences, n_sequences)
  """
  n_sequences <- nrow(sequences)
  dist_matrix <- matrix(0, nrow = n_sequences, ncol = n_sequences)
  
  for (i in 1:n_sequences) {
    for (j in i:n_sequences) {
      distance <- optimal_matching_distance(
        sequences[i, ], 
        sequences[j, ], 
        indel_cost = indel_cost, 
        substitution_cost = substitution_cost
      )
      dist_matrix[i, j] <- distance
      dist_matrix[j, i] <- distance
    }
  }
  
  return(dist_matrix)
}



# MAIN EXECUTION


# Load sequences
sequences <- as.matrix(read.csv("sequences_m1.csv", header = TRUE))

cat("=" %&% rep("=", 58), "\n")
cat("CALCUL DES DISTANCE MATRICES\n")
cat("=" %&% rep("=", 58), "\n")

# Hamming Distance
cat("\n1. HAMMING DISTANCE\n")
cat("-" %&% rep("-", 58), "\n")
dist_hamming <- hamming_distance_matrix(sequences)
cat("Hamming distance matrix shape:", dim(dist_hamming), "\n")
non_zero_hamming <- dist_hamming[dist_hamming > 0]
cat("Min distance:", min(non_zero_hamming), "\n")
cat("Max distance:", max(dist_hamming), "\n")
cat("Mean distance:", mean(non_zero_hamming), "\n")

# Optimal Matching
cat("\n2. OPTIMAL MATCHING DISTANCE\n")
cat("-" %&% rep("-", 58), "\n")
dist_om <- optimal_matching_matrix(sequences)
cat("Optimal Matching distance matrix shape:", dim(dist_om), "\n")
non_zero_om <- dist_om[dist_om > 0]
cat("Min distance:", min(non_zero_om), "\n")
cat("Max distance:", max(dist_om), "\n")
cat("Mean distance:", mean(non_zero_om), "\n")

# Save
cat("\n3. SAUVEGARDE\n")
cat("-" %&% rep("-", 58), "\n")
saveRDS(dist_hamming, "distance_matrix_hamming.rds")
saveRDS(dist_om, "distance_matrix_om.rds")
cat("✓ Distance matrices saved\n")
